lib/segment/src/index/field_index/full_text_index/inverted_index.rs

```diff
@@ -0,0 +1,193 @@
+use std::collections::{BTreeMap, BTreeSet};
+
+use serde::{Deserialize, Serialize};
+
+use crate::index::field_index::full_text_index::postings_iterator::intersect_btree_iterator;
+use crate::index::field_index::{CardinalityEstimation, PayloadBlockCondition, PrimaryCondition};
+use crate::types::{FieldCondition, Match, MatchText, PayloadKeyType, PointOffsetType};
+
+type PostingList = BTreeSet<PointOffsetType>;
+
+#[derive(Default, Serialize, Deserialize, Clone)]
+pub struct Document {
+    pub tokens: BTreeSet<String>,
+}
+
+impl Document {
+    pub fn is_empty(&self) -> bool {
+        self.tokens.is_empty()
+    }
+}
+
+pub struct ParsedQuery {
+    pub tokens: BTreeSet<String>,
+}
+
+impl ParsedQuery {
+    pub fn check_match(&self, document: &Document) -> bool {
+        // Check that all tokens are in document
+        self.tokens
+            .iter()
+            .all(|query_token| document.tokens.contains(query_token))
+    }
+}
+
+pub struct InvertedIndex {
+    postings: BTreeMap<String, PostingList>,
+    pub point_to_docs: Vec<Document>,
+    pub points_count: usize,
+}
+
+impl InvertedIndex {
+    pub fn new() -> InvertedIndex {
+        InvertedIndex {
+            postings: BTreeMap::new(),
+            point_to_docs: Vec::new(),
+            points_count: 0,
+        }
+    }
+
+    pub fn index_document(&mut self, idx: PointOffsetType, document: Document) {
```

Comment:
I propose to return from this `fn` when document has no tokens. Because in this struct we assume that empty tokens list == not initialized point

Reply:
Or keep as is with changing struct definition
`pub point_to_docs: Vec<Document>`
to
`pub point_to_docs: Vec<Option<Document>>,`

Reply:
:+1: